### PR TITLE
Fix gap in primitive circular meshes due to floating point error

### LINKED ISSUE #93865: Round primitive meshes contain gaps.
### Tested versions

v4.3.beta2.official [b75f0485b]

Tested version on v4.3.beta2.official [b75f0485b], but likely has been present for a long time. First I noticed it https://github.com/godotengine/godot/pull/80710#issuecomment-1690749054 (and subsequently forgot.)

### System information

Godot v4.3.beta2 - Windows 10.0.22631 - GLES3 (Compatibility) - NVIDIA GeForce RTX 4090 (NVIDIA; 31.0.15.3699) - AMD Ryzen 9 7950X 16-Core Processor (32 Threads)

### Issue description

Primitive meshes that contain round surfaces, i.e. SphereMesh, CapsuleMesh, CylinderMesh and TorusMesh, have minuscule gaps. This is due to the fact that some of the vertices at the seams do not align. The gaps are unlikely to be seen in regular use but here's an image of an unconventional shader that exaggerate them:

capsule_and_sphere

Capsule and sphere, using a shader that expands vertices based on their floating-point presentations mantissa (for visualization purposes only; do not use the shader to debug the issue, as it might be inaccurate):

```glsl
shader_type spatial;
render_mode cull_disabled, unshaded;

void vertex() {
	vec3 n = normalize(vec3(floatBitsToUint(VERTEX) & uvec3(255)) - 127.5);
	VERTEX += NORMAL * (dot(NORMAL, n) + 1.) * 0.05;
}

void fragment() {
	ALBEDO = FRONT_FACING ? vec3(0.) : vec3(1., 0., 1.);
}
```

The gaps are more of an issue when manipulating the meshes in code, e.g. when iterating unique vertices to create edge-meshes.

The cause is likely a precision issue caused by calculating the same position twice with different values. Example code from CylinderMesh:

https://github.com/godotengine/godot/blob/9db1a963beae8056cbd30d692d4160d09c10b2dc/scene/resources/3d/primitive_meshes.cpp#L1010-L1015

 (`sin(0)` and `sin(TAU)` aren't identical)

For a fix I suggest either reusing values for positions already calculated or ensuring that the values used to derive the positions are identical.

### Steps to reproduce

The unaligned vertices can be for example found using the following script:
```gdscript
extends Node

func _ready() -> void:
	print_gaps(SphereMesh.new())
	print_gaps(CapsuleMesh.new())
	print_gaps(CylinderMesh.new())
	print_gaps(TorusMesh.new())

func print_gaps(mesh: PrimitiveMesh) -> void:
	var vertex_array: PackedVector3Array = mesh.get_mesh_arrays()[Mesh.ArrayType.ARRAY_VERTEX]
	for i: int in vertex_array.size():
		var v1: Vector3 = vertex_array[i]
		for j: int in range(i+1, vertex_array.size()):
			var v2: Vector3 = vertex_array[j]
			if v1 != v2 and v1.is_equal_approx(v2):
				printt("Gap:", v1, v2)
```

### Minimal reproduction project (MRP)

Project containing both the shader used to demonstrate the gaps and the script to find them:
[primitive_meshes.zip](https://github.com/user-attachments/files/16069240/primitive_meshes.zip)

### PR DESCRIPTION
**Pull request by ayanchavand**
Fixes #93865 
This is my first engine contribution besides docs therefore I have only fixed the sphere mesh for now as I am sure there will be changes needed in my implementation.
- Added 4 functions including `custom_sin` , `custom_cos `, `remove_negative_zero_vector3 `and  `remove_negative_zero`
- Replaced sin and cos with custom functions for fixing the floating point error and used remove_negative_zero for standard 0 comparison 
image

### CODE REVIEW

File: scene/resources/3d/primitive_meshes.cpp [hunk 1]
```diff
@@ -41,6 +41,30 @@
...
+double custom_cos(double x) {
+    x = std::fmod(x, Math_TAU);
+    if (x < 0) x += Math_TAU;
+    return remove_negative_zero(std::cos(x));
+}
+
+Vector3 remove_negative_zero_vector3(const Vector3& v) {
+    return Vector3(remove_negative_zero(v.x),
+                   remove_negative_zero(v.y),
+                   remove_negative_zero(v.z));
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:65]:
```suggestion
static double _remove_negative_zero(double p_value) {
	return Math::abs(p_value) < EPSILON ? 0.0 : p_value;
}

static double _custom_sin(double p_x) {
	p_x = Math::fmod(p_x, Math_TAU);
	if (p_x < 0) {
		p_x += Math_TAU;
	}
	return _remove_negative_zero(Math::sin(p_x));
}

static double _custom_cos(double p_x) {
	p_x = Math::fmod(p_x, Math_TAU);
	if (p_x < 0) {
		p_x += Math_TAU;
	}
	return _remove_negative_zero(Math::cos(p_x));
}

static Vector3 _remove_negative_zero_vector3(const Vector3 &p_v) {
	return Vector3(_remove_negative_zero(p_v.x), _remove_negative_zero(p_v.y), _remove_negative_zero(p_v.z));
```
Use built-in methods, and tabs for indentation, and style for static methods

File: scene/resources/3d/primitive_meshes.cpp [hunk 2]
```diff
@@ -966,6 +990,8 @@ void CylinderMesh::_create_mesh_array(Array &p_arr) const {
 void CylinderMesh::create_mesh_array(Array &p_arr, float top_radius, float bottom_radius, float height, int radial_segments, int rings, bool cap_top, bool cap_bottom, bool p_add_uv2, const float p_uv2_padding) {
 	int i, j, prevrow, thisrow, point;
 	float x, y, z, u, v, radius, radius_h;
+	const double EPSILON = 1e-10;
+
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:994]:
```suggestion
	const double EPSILON = 1e-10;
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 3]
```diff
@@ -1011,8 +1037,8 @@ void CylinderMesh::create_mesh_array(Array &p_arr, float top_radius, float botto
 			u = i;
 			u /= radial_segments;
 
-			x = sin(u * Math_TAU);
-			z = cos(u * Math_TAU);
+			x = custom_sin(u * Math_TAU);
+			z = custom_cos(u * Math_TAU);
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1041]:
```suggestion
			x = _custom_sin(u * Math_TAU);
			z = _custom_cos(u * Math_TAU);
```

File: scene/resources/3d/primitive_meshes.cpp [hunk 4]
```diff
@@ -1813,32 +1839,34 @@ void SphereMesh::create_mesh_array(Array &p_arr, float radius, float height, int
...
+            } else {
+                p = remove_negative_zero_vector3(Vector3(x * radius * w, y, z * radius * w));
+                normal = remove_negative_zero_vector3(Vector3(x * w * scale, radius * (y / scale), z * w * scale));
+                normal.normalize();
+            }
+
+            points.push_back(p);
+            normals.push_back(normal);
+            ADD_TANGENT(remove_negative_zero(z), 0.0, remove_negative_zero(-x), 1.0)
+            uvs.push_back(Vector2(u, v));
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1869]:
```suggestion
	for (j = 0; j <= (rings + 1); j++) {
		float v = static_cast<float>(j) / (rings + 1);
		float phi = M_PI * v;
		float w = _custom_sin(phi);
		y = scale * _custom_cos(phi);

		for (i = 0; i <= radial_segments; i++) {
			float u = static_cast<float>(i) / radial_segments;
			float theta = u * Math_TAU;
			x = _custom_sin(theta);
			z = _custom_cos(theta);

			Vector3 p;
			Vector3 normal;

			if (is_hemisphere && y < 0.0) {
				p = _remove_negative_zero_vector3(Vector3(x * radius * w, 0.0, z * radius * w));
				normal = Vector3(0.0, -1.0, 0.0);
			} else {
				p = _remove_negative_zero_vector3(Vector3(x * radius * w, y, z * radius * w));
				normal = _remove_negative_zero_vector3(Vector3(x * w * scale, radius * (y / scale), z * w * scale));
				normal.normalize();
			}

			points.push_back(p);
			normals.push_back(normal);
			ADD_TANGENT(_remove_negative_zero(z), 0.0, _remove_negative_zero(-x), 1.0)
			uvs.push_back(Vector2(u, v));
```

**Review by AThousandShips — DISMISSED** [review]

File: scene/resources/3d/primitive_meshes.cpp (hunk 2 above)
Comment (AThousandShips) [reply scene/resources/3d/primitive_meshes.cpp:996]: Also this isn't used anywhere is it?

Comment (ayanchavand) [reply scene/resources/3d/primitive_meshes.cpp:996]: no, I forgot to remove it when testing for a fix.

**Comment by AThousandShips** [comment]
It would be good to add some unit tests for this in `tests/scene/test_primitives.h` to make sure it works correctly

**Comment by ayanchavand** [comment]
I have never tried unit testing before, can you share some resources as to how I can implement it for this case?

**Comment by AThousandShips** [comment]
See [this](https://docs.godotengine.org/en/latest/contributing/development/core_and_modules/unit_testing.html) and you can look at the existing code to learn how it's used

File: scene/resources/3d/primitive_meshes.cpp [hunk 5]
```diff
@@ -1814,30 +1840,32 @@ void SphereMesh::create_mesh_array(Array &p_arr, float radius, float height, int
...
 	prevrow = 0;
 	for (j = 0; j <= (rings + 1); j++) {
-		float v = j;
-		float w;
-
-		v /= (rings + 1);
-		w = sin(Math_PI * v);
-		y = scale * cos(Math_PI * v);
+		float v = static_cast<float>(j) / (rings + 1);
+		float phi = M_PI * v;
```
Comment (AThousandShips) [inline scene/resources/3d/primitive_meshes.cpp:1844]:
```suggestion
		float phi = Math_PI * v;
```

**Comment by Lielay9** [comment]
The Torus and Capsule still have gaps (since nothing was done to them) and the cylinder still has gaps in the rings of the caps. 

Additionally the approach taken is bit worrisome, as there is at least a theoretical possibility that vertices that should be different will be snapped to the same position if they're close enough (difference is less than epsilon). The sin and cos functions were fine, the problem was with the last vertex in the ring not matching the first; simpler approach would be to make the last iterations position zero with an if statement (whilst ensuring the UV:s remain correct).

**Comment by fire** [comment]
Hi, I also work on CSG, would it be ok to also test CSGMesh3D too?

**Comment by fire** [comment]
Superseded by: https://github.com/godotengine/godot/pull/100020

**Pull request closed without merging**